File: python/ch01_environment/check_connection.py
import os
import sys
from typing import Dict

from kafka import KafkaAdminClient


def load_env() -> Dict[str, str]:
    """Load Kafka connection details from env and validate required fields.

    This mirrors the Bash env loader logic at a high level.
    """
    bootstrap = os.getenv("KAFKA_BOOTSTRAP_SERVERS")
    if not bootstrap:
        print(
            "[error] KAFKA_BOOTSTRAP_SERVERS is not set. "
            "Copy infra/env-example.msk or infra/env-example.local and source it before running.",
            file=sys.stderr,
        )
        sys.exit(1)

    security_protocol = os.getenv("KAFKA_SECURITY_PROTOCOL", "PLAINTEXT")
    sasl_mechanism = os.getenv("KAFKA_SASL_MECHANISM")
    sasl_username = os.getenv("KAFKA_SASL_USERNAME")
    sasl_password = os.getenv("KAFKA_SASL_PASSWORD")

    return {
        "bootstrap_servers": bootstrap.split(","),
        "security_protocol": security_protocol,
        "sasl_mechanism": sasl_mechanism,
        "sasl_plain_username": sasl_username,
        "sasl_plain_password": sasl_password,
    }


def check_connection() -> None:
    config = load_env()

    print(f"[info] Checking connectivity to Kafka at {','.join(config['bootstrap_servers'])}...")

    # Filter out empty auth values so local PLAINTEXT works without extra config.
    admin_kwargs = {k: v for k, v in config.items() if v}

    try:
        client = KafkaAdminClient(**admin_kwargs, client_id="kafka-tutorials-python-check")
        topics = client.list_topics()
        print(f"[info] Successfully connected. Found {len(topics)} topics.")
    except Exception as exc:  # pragma: no cover - simple demo script
        print("[error] Failed to connect to Kafka:", file=sys.stderr)
        print(f"        {exc}", file=sys.stderr)
        sys.exit(1)
    finally:
        try:
            client.close()  # type: ignore[has-type]
        except Exception:
            pass


if __name__ == "__main__":
    check_connection()

File: java/ch01_environment/CheckConnection.java
import java.util.Properties;

import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.ListTopicsResult;

public class CheckConnection {
    public static void main(String[] args) throws Exception {
        String bootstrap = System.getenv("KAFKA_BOOTSTRAP_SERVERS");
        if (bootstrap == null || bootstrap.isEmpty()) {
            System.err.println("[error] KAFKA_BOOTSTRAP_SERVERS is not set. " +
                    "Copy infra/env-example.msk or infra/env-example.local and source it before running.");
            System.exit(1);
        }

        String securityProtocol = getenvOrDefault("KAFKA_SECURITY_PROTOCOL", "PLAINTEXT");
        String saslMechanism = getenvOrDefault("KAFKA_SASL_MECHANISM", "");
        String saslUsername = getenvOrDefault("KAFKA_SASL_USERNAME", "");
        String saslPassword = getenvOrDefault("KAFKA_SASL_PASSWORD", "");

        Properties props = new Properties();
        props.put("bootstrap.servers", bootstrap);
        props.put("client.id", "kafka-tutorials-java-check");
        props.put("security.protocol", securityProtocol);

        if (!saslMechanism.isEmpty()) {
            props.put("sasl.mechanism", saslMechanism);
        }
        if (!saslUsername.isEmpty() && !saslPassword.isEmpty()) {
            props.put("sasl.jaas.config", String.format(
                    "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"%s\" password=\"%s\";",
                    saslUsername,
                    saslPassword
            ));
        }

        System.out.println("[info] Checking connectivity to Kafka at " + bootstrap + "...");

        try (AdminClient admin = AdminClient.create(props)) {
            ListTopicsResult result = admin.listTopics();
            int count = result.names().get().size();
            System.out.println("[info] Successfully connected. Found " + count + " topics.");
        } catch (Exception e) {
            System.err.println("[error] Failed to connect to Kafka:");
            System.err.println("        " + e.getMessage());
            System.exit(1);
        }
    }

    private static String getenvOrDefault(String key, String defaultValue) {
        String value = System.getenv(key);
        return value != null ? value : defaultValue;
    }
}
